### PR TITLE
:sparkles: feat: support moving capi resources to a desired namespace

### PR DESCRIPTION
**Pull request by takirala**
This change supports moving capi clusters from namespace A on one cluster to namespace B on another cluster. Previously, clusters will always be relocated to same namespace as they originate from. 

This was done in interest of making it possible to relocate `Cluster` resources from `default` namespace to a non default namespace (`workspace` namespace in context of kommander) to keep resources more _organized_.

This change is being tested in https://github.com/mesosphere/kommander/pull/2787

### CODE REVIEW

File: cmd/clusterctl/client/cluster/mover.go
Comment (dlipovetsky) [inline cmd/clusterctl/client/cluster/mover.go]: Strictly speaking, you don't need to consider the namespace of any "reference" fields. Although they are of the v1.ObjectReference go type, the namespace field is not used. https://github.com/kubernetes-sigs/cluster-api/issues/6539 describes the work in progress to use a different go type. However, while that work is not merged, it may be important to respect the namespace fields.

Comment (takirala) [reply cmd/clusterctl/client/cluster/mover.go]: thanks for linking to that upstream issue. 

but yeah i agree, until that can be merged we need to update the fields, cuz otherwise admisison webhook is failing the `Cluster` creation saying `.metadata.namespace != spec.someRef.namespace`

**Comment by takirala** [comment]
@dlipovetsky @dkoshkin what do you think of reviewing this and merging this while we wait for upstream contrib to be reviewed/merged/released !?

**Comment by dlipovetsky** [comment]
@takirala will you update this with commits from https://github.com/kubernetes-sigs/cluster-api/pull/7966, or open a new PR?

**Comment by takirala** [comment]
Synced this PR with current state of kubernetes-sigs/cluster-api#7966

**Comment by takirala** [comment]
@dlipovetsky @dkoshkin 

Will we have a release of `mesosphere/cluster-api` before 2.5 ? (So that i can use this change in kommander 2.5.0)

**Comment by dlipovetsky** [comment]
> @dlipovetsky @dkoshkin
> 
> Will we have a release of `mesosphere/cluster-api` before 2.5 ? (So that i can use this change in kommander 2.5.0)

Yes :smile:  It's in this sprint: https://d2iq.atlassian.net/browse/D2IQ-95558

Thanks for keeping this in sync with the upstream PR!

File: cmd/clusterctl/client/cluster/mover.go
Comment (dkoshkin) [inline cmd/clusterctl/client/cluster/mover.go]: Should we also change this to an optional var to keep the existing calls the same?
```suggestion
func setClusterPause(proxy Proxy, clusters []*node, value bool, dryRun bool, mutators ...ResourceMutatorFunc) error {
```

File: cmd/clusterctl/client/cluster/mover.go
Comment (dkoshkin) [inline cmd/clusterctl/client/cluster/mover.go]: To make it consistent.
```suggestion
func (o *objectMover) move(graph *objectGraph, toProxy Proxy, mutators ...ResourceMutatorFunc) error {
```

File: cmd/clusterctl/client/move_test.go
Comment (dkoshkin) [inline cmd/clusterctl/client/move_test.go]: Please revert

**Comment by dlipovetsky** [comment]
Thanks for the good feedback @dkoshkin!

**Comment by dkoshkin** [comment]
@takirala Please also rebase against https://github.com/mesosphere/cluster-api/tree/d2iq/release-1.3.3 and change the base branch here 🙏

**Review by dkoshkin — APPROVED** [review]
Thanks for all the effort here, looks great!

**Comment by dlipovetsky** [comment]
`make test` is happy locally

**Review by dlipovetsky — APPROVED** [review]

**Comment by takirala** [comment]
Synced changes from this PR to upstream PR - kubernetes-sigs/cluster-api#7966

**Pull request merged**
